sentry/silkworm/sentry/common/task_group.cpp

```diff
@@ -0,0 +1,125 @@
+/*
+   Copyright 2022 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#include "task_group.hpp"
+
+#include <cassert>
+#include <tuple>
+#include <utility>
+
+#include <boost/asio/bind_cancellation_slot.hpp>
+#include <boost/asio/co_spawn.hpp>
+#include <boost/asio/this_coro.hpp>
+#include <boost/system/errc.hpp>
+#include <boost/system/system_error.hpp>
+
+#include <silkworm/common/log.hpp>
+
+namespace silkworm::sentry::common {
+
+using namespace boost::asio;
+
+void TaskGroup::spawn(any_io_executor&& executor, awaitable<void> task) {
+    std::scoped_lock lock(mutex_);
+
+    if (is_closed_) {
+        throw SpawnAfterCloseError();
+    }
+
+    auto task_id = ++last_task_id_;
+
+    auto [it, ok] = tasks_.emplace(
+        std::piecewise_construct,
+        std::forward_as_tuple(task_id),
+        std::forward_as_tuple());
+    assert(ok);
```

Comment:
Just in case you need it, you can use `SILKWORM_ASSERT` instead of `assert` to have the condition checked also in release builds